### ISSUE TITLE
Handling of units in CCDData instantiation

### ISSUE THREAD

**Issue by mwcraig** [astropy/astropy#7619]
The issue of how permissive to make `nddata.CCDData` in opening images came up in #7608, in which opening an HST image failed because the `BUNIT` in the header couldn't be processed as a valid unit.

I'm opening this issue to start a broader discussion about how CCDData should behave in the future.

Right now creating a `CCDData` object fails if there is no unit (either explicitly given in the constructor or in the `BUNIT` of a FITS file).

In #7608, @astrofrog suggested the class should be more permissive:

> I think more generally, we've raised the idea in the past of being more forgiving on input and strict only on output. This seems like a good example of why. I think this should just read in an if the user accesses the unit it would return an unrecognizedunit('ELECTRONS')

To focus this a bit:

1. If there is no unit in the header (if creating from a file) and no unit in the constructor,  should construction succeed or fail?  If it succeeds it would be assigned a unit of `u.dimensionless_unscaled`.

2. If there is a unit in the header and the unit does not parse properly, what should the result be? Fail with an error, proceed with an `unrecognizedunit` (maybe with a warning), something else?

Any of these would be straightforward to implement, and it ab easy to modify ccdproc to handle any of these changes (ccdproc currently requires that there be a unit on CCDData), input on this would be very helpful...

**Comment by MSeifert04** [comment]
I'm really not sure. The question is what a user expects and what would be easiest to avoid or fix manually.

If we carry around an `UnrecognizedUnit` and it fails several function calls later because something expects a valid unit then it's probably not very helpful for the user - because he has to investigate where the UnrecognizedUnit came from and then figure out how to make it a valid unit or how it can be avoided altogether.

Dropping the unit because it didn't parse correctly also seems wrong because the behavior then becomes "seemingly" unpredictable.

Maybe a warning instead of an exception would be better, but to be honest: Who really looks at warnings? If I'm working with fits files and nddata there are typically so many warnings/infos/stuff that I can understand any user who hides them completely. But you cannot hide an exception.

From my personal experience: Failing early with a good exception message is easy for a user to fix, having something "trying to guess what the user intended" is more complicated. At least that's something I found really annoying in JavaScript (and to some extend Python 2).

**Comment by bsipocz** [comment]
Can we both have the cake and eat it? A nice explicit error message detailing out what the `UnrecognizedUnit` was, and mention how to feed in a valid one via the `unit` keyword? 

Unfortunately HST is not the only big archive not to follow the standards, but still, building a custom parsing of `electrons--> electron`, `angstroms --> angstrom`, etc, I feel we would contradict ourselves.

**Comment by bsipocz** [comment]
@pllim - Nice. You don't have that many exceptions though, it could be much worse :trollface:

**Comment by bsipocz** [comment]
Has there been any decisions on this? If not, maybe this is something to discuss at the meeting, or close the issue?

**Comment by mwcraig** [comment]
No, let's defer to the meeting...agreed that failing early and explicitly is better.

**Comment by pllim** [comment]
@mwcraig , I added it to potentials things to "breakout" about at the meeting. FYI.
